onmt/inputters/item.py

```diff
@@ -0,0 +1,280 @@
+         'align': alignment sparse tensor
+        }
+    """
+    tbatch = {}
+    tbatchsrc = [torch.LongTensor(ex['src']['src_ids']) for ex in minibatch]
+    padidx = vocabs['src'][DefaultTokens.PAD]
+    tbatchsrc = pad_sequence(tbatchsrc, padding_value=padidx)
+    if len(minibatch[0]['src'].keys()) > 2:
+        tbatchfs = [tbatchsrc]
+        for feat in minibatch[0]['src'].keys():
+            if feat not in ['src', 'src_ids']:
+                tbatchfeat = [torch.LongTensor(ex['src'][feat])
+                              for ex in minibatch]
+                padidx = vocabs['src_feats'][feat][DefaultTokens.PAD]
+                tbatchfeat = pad_sequence(tbatchfeat, padding_value=padidx)
+                tbatchfs.append(tbatchfeat)
+        tbatchsrc = torch.stack(tbatchfs, dim=2)
+    else:
+        tbatchsrc = tbatchsrc[:, :, None]
+    # Need to add features in last dimensions
+
+    tbatch['src'] = tbatchsrc
+    tbatch['indices'] = torch.LongTensor([ex['indices'] for ex in minibatch])
+    tbatch['srclen'] = torch.LongTensor([len(ex['src']['src_ids'])
+                                         for ex in minibatch])
+
+    if minibatch[0]['tgt'] is not None:
+        tbatchtgt = [torch.LongTensor(ex['tgt']['tgt_ids'])
+                     for ex in minibatch]
+        padidx = vocabs['tgt'][DefaultTokens.PAD]
+        tbatchtgt = pad_sequence(tbatchtgt, padding_value=padidx)
+        tbatchtgt = tbatchtgt[:, :, None]
+        tbatchtgtlen = torch.LongTensor([len(ex['tgt']['tgt_ids'])
+                                         for ex in minibatch])
+        tbatch['tgt'] = tbatchtgt
+        tbatch['tgtlen'] = tbatchtgtlen
+
+    if 'align' in minibatch[0].keys() and minibatch[0]['align'] is not None:
+        sparse_idx = []
+        for i, ex in enumerate(minibatch):
+            for src, tgt in parse_align_idx(ex['align']):
+                sparse_idx.append([i, tgt + 1, src])
+        tbatchalign = torch.LongTensor(sparse_idx)
+        tbatch['align'] = tbatchalign
+
+    if 'src_map' in minibatch[0].keys():
+        src_vocab_size = max([max(ex['src_map']) for ex in minibatch]) + 1
+        src_map = torch.zeros(tbatchsrc.size(0),
+                              len(tbatch['srclen']),
+                              src_vocab_size)
+        for i, ex in enumerate(minibatch):
+            for j, t in enumerate(ex['src_map']):
+                src_map[j, i, t] = 1
+        tbatch['src_map'] = src_map
+
+    if 'alignment' in minibatch[0].keys():
+        alignment = torch.zeros(tbatchtgt.size(0),
+                                len(tbatch['srclen'])).long()
+        for i, ex in enumerate(minibatch):
+            alignment[:len(ex['alignment']), i] = \
+                torch.LongTensor(ex['alignment'])
+        tbatch['alignment'] = alignment
+
+    if 'src_ex_vocab' in minibatch[0].keys():
+        tbatch['src_ex_vocab'] = [ex['src_ex_vocab']
+                                  for ex in minibatch]
+    return tbatch
```

Comment:
A few comments around  here would not hurt.

Reply:
the doc string is not enough ?